File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/ai/harness/maven/RequireEnvShebangUnderRule.java
package ai.harness.maven;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Collection;
import java.util.stream.Stream;

/**
 * Rule that requires executable scripts to use /usr/bin/env shebang.
 */
public class RequireEnvShebangUnderRule implements HarnessCheckRule {
    private static final String CATEGORY = "requireEnvShebangUnder";

    @Override
    public boolean applies(JsonNode manifest) {
        return HarnessCheckHelper.applies(manifest, CATEGORY);
    }

    @Override
    public Collection<Finding> validate(Path root, JsonNode manifest) throws MojoExecutionException {
        JsonNode catNode = manifest.get(CATEGORY);
        List<String> directories = HarnessCheckHelper.extractPaths(catNode.get("parameters").get("directories"));
        String expectedPrefix = catNode.get("parameters").get("expectedPrefix").asText();
        String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        return directories.stream()
                .flatMap(dir -> {
                    try {
                        Path dirPath = root.resolve(dir);
                        return HarnessCheckHelper.safeFileOrWalk(root, dirPath).stream()
                                .filter(Files::isExecutable)
                                .flatMap(file -> validateShebang(root, file, expectedPrefix, severity).stream());
                    } catch (MojoExecutionException e) {
                        return Stream.empty();
                    }
                })
                .toList();
    }

    private List<Finding> validateShebang(Path root, Path file, String expectedPrefix, String severity) throws MojoExecutionException {
        String text = HarnessCheckHelper.readFile(root, file);
        return text.startsWith("#!") && !text.startsWith(expectedPrefix)
                ? List.of(new Finding(severity, CATEGORY, "executable script should use /usr/bin/env shebang: " + root.relativize(file)))
                : List.of();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/ai/harness/gradle/RequireEnvShebangUnderRule.kt
package ai.harness.gradle

import kotlinx.serialization.json.JsonObject
import java.nio.file.Path
import kotlin.io.path.div
import kotlin.io.path.isDirectory
import kotlin.io.path.isExecutable
import kotlin.io.path.readText
import kotlin.io.path.relativeTo

/**
 * Rule that requires executable scripts under certain directories to use /usr/bin/env shebang.
 */
class RequireEnvShebangUnderRule : HarnessCheckRule {
	override fun applies(manifest: JsonObject): Boolean {
		val category = "requireEnvShebangUnder"
		val catObj = manifest[category]?.jsonObject ?: return false
		val enabled = catObj["enabled"]?.jsonPrimitive?.contentOrNull?.toBoolean() ?: true
		return enabled
	}

	override fun validate(manifest: JsonObject, root: Path, psiResults: HarnessPsiResults?): Collection<Finding> {
		val category = "requireEnvShebangUnder"
		val severity = HarnessCheck.Companion.severityOf(manifest, category)
		val catObj = manifest[category]?.jsonObject ?: return emptyList()
		val parametersObj = catObj["parameters"]?.jsonObject ?: return emptyList()
		val messagesObj = catObj["messages"]?.jsonObject ?: return emptyList()
		val directories = HarnessCheck.Companion.stringArrayFrom(parametersObj, "directories")
		val expectedPrefix = HarnessCheck.Companion.stringFrom(parametersObj, "expectedPrefix")

		return buildSet<Finding> {
			directories.forEach { dirPath ->
				val dir = root / dirPath
				if (!dir.isDirectory()) {
					return@forEach
				}

				val (files, _) = HarnessCheck.Companion.walkSafe(root, dir)
				files.filter { it.isExecutable() }.forEach { file ->
					val firstLine = file.readText().lineSequence().firstOrNull() ?: ""
					if (firstLine.startsWith("#!") && !firstLine.startsWith(expectedPrefix)) {
						val msg = HarnessCheck.Companion.stringFrom(messagesObj, "default").takeIf { it.isNotEmpty() } ?: "executable script should use /usr/bin/env shebang: ${file.relativeTo(root)}"
						add(Finding(severity, category, msg))
					}
				}
			}
		}.toList()
	}
}
